lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,726 @@
+    #   DWORD dwCheckPoint;
+    #   DWORD dwWaitHint;
+    # } SERVICE_STATUS, *LPSERVICE_STATUS;
+    class SERVICE_STATUS < FFI::Struct
+      layout(
+        :dwServiceType, :dword,
+        :dwCurrentState, :dword,
+        :dwControlsAccepted, :dword,
+        :dwWin32ExitCode, :dword,
+        :dwServiceSpecificExitCode, :dword,
+        :dwCheckPoint, :dword,
+        :dwWaitHint, :dword,
+      )
+    end
+
+    # typedef struct _QUERY_SERVICE_CONFIGW {
+    #   DWORD  dwServiceType;
+    #   DWORD  dwStartType;
+    #   DWORD  dwErrorControl;
+    #   LPWSTR lpBinaryPathName;
+    #   LPWSTR lpLoadOrderGroup;
+    #   DWORD  dwTagId;
+    #   LPWSTR lpDependencies;
+    #   LPWSTR lpServiceStartName;
+    #   LPWSTR lpDisplayName;
+    # } QUERY_SERVICE_CONFIGW, *LPQUERY_SERVICE_CONFIGW;
+    class QUERY_SERVICE_CONFIGW < FFI::Struct
+      layout(
+        :dwServiceType, :dword,
+        :dwStartType, :dword,
+        :dwErrorControl, :dword,
+        :lpBinaryPathName, :pointer,
+        :lpLoadOrderGroup, :pointer,
+        :dwTagId, :dword,
+        :lpDependencies, :pointer,
+        :lpServiceStartName, :pointer,
+        :lpDisplayName, :pointer,
+      )
+    end
+
+    # Start a windows service, assume that the service is already in the stopped state
+    #
+    # @param [:string] service_name name of the service to start
+    def start(service_name)
+      open_service(service_name, SC_MANAGER_CONNECT, SERVICE_START | SERVICE_QUERY_STATUS) do |service|
+        # don't attempt to fail here if the service isn't stopped because windows error codes
+        # are likely more informative than ours and a failed call to StartServiceW will produce
+        # those errors
+        wait_for_pending_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
+        if StartServiceW(service, 0, nil) == FFI::WIN32_FALSE
+          raise Puppet::Util::Windows::Error.new(_("Failed to start the service"))
+        end
+        unless wait_for_pending_transition(service, SERVICE_START_PENDING, SERVICE_RUNNING)
+          raise Puppet::Error.new(_("Failed to start the service, after calling StartService the service is not in SERVICE_START_PENDING or SERVICE_RUNNING"))
+        end
+      end
+    end
+    module_function :start
+
+    # Use ControlService to send a stop signal to a windows service
+    #
+    # @param [:string] service_name name of the service to stop
+    def stop(service_name)
+      FFI::MemoryPointer.new(SERVICE_STATUS.size) do |status_ptr|
+        status = SERVICE_STATUS.new(status_ptr)
+        open_service(service_name, SC_MANAGER_CONNECT, SERVICE_STOP | SERVICE_QUERY_STATUS) do |service|
+          # don't attempt to fail here if the service isn't started because windows error codes
+          # are likely more informative than ours and a failed call to ControlService will produce
+          # those errors
+          wait_for_pending_transition(service, SERVICE_START_PENDING, SERVICE_RUNNING)
```

Comment:
Is it necessary to wait for the pending transition? The [state transition chart](https://docs.microsoft.com/en-us/windows/desktop/services/service-status-transitions) seems to indicate that you can stop a service from any state so if this errors, that's something with the service (and not on our end). Plus, we'll need to wait for `CONTINUE_PENDING` anyways as this currently stands, so it'd be nice if it could be simplified.

Reply:
Yeah, it is, see https://github.com/puppetlabs/puppet/pull/7017#issuecomment-417805844